File: 2024/src/code_19.py
from copy import deepcopy
from itertools import chain


def try_build_pattern(pattern: str, towels: list[str]) -> list[list[str]]:
    def towel_to_patterns(towel: str):
        if len(towel) >= len(pattern):
            return [[towel]]
        else:
            return [
                [towel, *ts] for ts in try_build_pattern(pattern[len(towel) :], towels)
            ]

    print("build", pattern, towels)
    usable_towels = [
        towel_to_patterns(towel) for towel in towels if pattern.startswith(towel)
    ]

    return list(chain.from_iterable(usable_towels))


def run():
    with open("19.inp") as rf:
        data = rf.read()

    towels = data.splitlines()[0].split(", ")

    patterns = data.splitlines()[2:]

    base_towels = set()
    for towel in towels:
        new_towels = deepcopy(towels)
        new_towels.remove(towel)

        buildable_from = try_build_pattern(towel, new_towels)

        if not buildable_from:
            base_towels.add(towel)
        else:
            for ts in buildable_from:
                for t in ts:
                    base_towels.add(t)

    doable = 0
    total = 0

    for pattern in patterns:
        n_patterns = len(try_build_pattern(pattern, base_towels))
        total += n_patterns

        print("done", pattern)
        if n_patterns > 0:
            doable += 1

    print("doable patterns: ", doable)

    print("new towels: ", len(base_towels), len(towels))


run()

File: 2024/src/code_19.rs
use itertools::Itertools;

fn build_pattern(pattern: &str, towels: &Vec<String>) -> i32 {
    let usable_towels = towels.iter().filter(|towel| pattern.starts_with(*towel));

    let possible_options = usable_towels.map(|towel| {
        if towel.len() >= pattern.len() {
            1
        } else {
            build_pattern(&pattern[towel.len()..], towels)
        }
    });

    possible_options.sum()
}

fn parse_towels(inp: &str) -> Vec<&str> {
    inp.split(", ").map(|v| v.to_owned()).collect_vec()
}

fn try_build_pattern<'a>(pattern: &str, towels: &Vec<&'a str>) -> Vec<Vec<&'a str>> {
    let usable_towels = towels
        .into_iter()
        .filter(|towel| pattern.starts_with(*towel));

    let possible_options = usable_towels.flat_map(|towel| {
        if towel.len() >= pattern.len() {
            vec![vec![*towel]]
        } else {
            try_build_pattern(&pattern[towel.len()..], towels)
                .into_iter()
                .map(|mut ts| {
                    ts.insert(0, towel);
                    ts
                })
                .collect_vec()
        }
    });

    possible_options.collect_vec()
}

pub fn run(data: &String) {
    let towels = data
        .lines()
        .next()
        .unwrap()
        .split(", ")
        .map(|v| v)
        .collect_vec();
    let patterns = data.lines().skip(2).map(|v| v.to_owned()).collect_vec();

    let mut base_towels = Vec::new();
    for (i, towel) in towels.iter().enumerate() {
        let mut new_towels = towels.clone();
        new_towels.remove(i);
        let buildable_from = try_build_pattern(towel, &new_towels);

        if buildable_from.is_empty() {
            base_towels.push(towel);
        } else {
            for ts in buildable_from {
                for t in ts {
                    base_towels.push(t);
                }
            }
        }
    }

    base_towels.sort();
    base_towels.dedup();
    println!("new towels: {:?}", base_towels);

    // let mut doable = 0;
    // let mut total = 0;
    // for pattern in patterns {
    //     let n_patterns = build_pattern(&pattern, &towels);
    //     total += n_patterns;
    //     println!("{:?} done", pattern);
    //     if n_patterns > 0 {
    //         doable += 1;
    //     }
    // }

    // println!("doable patterns: {:?}", doable);
}
